core/src/dsnp/graph_page.rs

```diff
@@ -100,32 +277,322 @@ impl GraphPage {
-	pub fn set_pages(&mut self, pages: HashMap<PageId, GraphPage>) {
+	pub fn set_pages(&mut self, pages: PageMap) {
 		self.pages = pages;
 	}
 
+	/// Get next available PageId for this graph
+	pub fn get_next_available_page_id(&self) -> Option<PageId> {
+		const MAX_PAGE_ID: PageId = 16; // todo: move this to config
+		let existing_pages =
+			self.pages.iter().map(|(page_id, _)| *page_id).collect::<HashSet<PageId>>();
+		for pid in 0..MAX_PAGE_ID {
+			if !existing_pages.contains(&pid) {
+				return Some(pid)
+			}
+		}
+
+		None
+	}
+
+	/// Remove all pages from this graph
+	pub fn clear(&mut self) {
+		self.pages.clear();
+	}
+
+	/// Import bundle of pages as a Public Graph
+	pub fn import_public<E: EncryptionBehavior>(
+		&mut self,
+		ImportBundle::<E> { connection_type, pages, .. }: ImportBundle<E>,
+	) -> Result<()> {
+		if connection_type != self.connection_type {
+			return Err(Error::msg("Incorrect connection type for graph import"))
+		}
+
+		let mut page_map = HashMap::<PageId, GraphPage>::new();
+		for page in pages.iter() {
+			match GraphPage::try_from(page) {
+				Err(e) => return Err(e),
+				Ok(p) => {
+					page_map.insert(page.page_id, p);
+				},
+			};
+		}
+
+		self.set_pages(page_map);
+
+		Ok(())
+	}
+
+	/// Import bundle of pages as a Private Graph
+	pub fn import_private<E: EncryptionBehavior>(
+		&mut self,
+		ImportBundle::<E> { connection_type, pages, keys, .. }: ImportBundle<E>,
+	) -> Result<()> {
+		if connection_type != self.connection_type {
+			return Err(Error::msg("Incorrect connection type for graph import"))
+		}
+
+		let mut page_map = PageMap::new();
+		for page in pages.iter() {
+			match GraphPage::try_from((page, &keys)) {
+				Err(e) => return Err(e),
+				Ok(p) => {
+					page_map.insert(page.page_id, p);
+				},
+			};
+		}
+
+		self.set_pages(page_map);
+
+		Ok(())
+	}
+
+	/// Import bundle of pages as a Private Graph, but without decrypting the encrypted portion
+	/// Useful for importing a foreign user's graph for inspection of PRIds
+	pub fn import_opaque<E: EncryptionBehavior>(
+		&mut self,
+		ImportBundle::<E> { connection_type, pages, .. }: ImportBundle<E>,
+	) -> Result<()> {
+		if connection_type.privacy_type() != PrivacyType::Private ||
+			self.connection_type.privacy_type() != PrivacyType::Private
+		{
+			return Err(Error::msg("Invalid privacy type for opaque Private graph import"))
+		}
+
+		let mut page_map = PageMap::new();
+		for page in pages.iter() {
+			match DsnpUserPrivateGraphChunk::try_from(page.payload.as_slice()) {
+				Err(e) => return Err(e.into()),
+				Ok(chunk) => {
+					page_map.insert(
+						page.page_id,
+						GraphPage {
+							page_id: page.page_id,
+							privacy_type: self.connection_type.privacy_type(),
+							content_hash: page.content_hash,
+							prids: chunk.prids,
+							connections: Vec::new(),
+						},
+					);
+				},
+			};
+		}
+
+		self.set_pages(page_map);
+
+		Ok(())
+	}
+
+	pub fn calculate_updates<E: EncryptionBehavior>(
```

Comment:
We could have implemented this in another PR but since you already done it then no complains.